### PR TITLE
Simplify Function::eval interface

### PR DESCRIPTION
**Pull request by garth-wells**
This changes the `Function` `eval` interface to take (i) a list of points and (ii) a list of cells indices that contain each point. Points with negative cell indices are ignored (the bounding box search now returns a negative index when the cell containing a point is not found). This makes point evaluations in parallel much simpler.

The change makes what is happening behind `eval` explicit and removes a ton of opaque magic behind the scenes, e.g. cell searches, assumptions on what to do when a point is 'close' to a cell but not within it, etc. Now allowing an arbitrary number of points to be processed permits a number of performance optimisations and minimises dynamics memory allocations.

Example usage:
```
x = numpy.array(......)
tree = geometry.BoundingBoxTree(mesh, mesh.geometry.dim)
cells = geometry.compute_first_entity_collision(tree, mesh, x)
f = u.eval(x, cells)
```

The is scope for a further optimisation is the user passes in points that are groups by the cell that contains them.

### CODE REVIEW

File: cpp/dolfin/function/Function.cpp [hunk 1]
```diff
@@ -266,42 +202,65 @@ void Function::eval(
         "fem::CoordinateMapping has not been attached to mesh.");
   }
 
-  std::size_t num_points = x.rows();
-  std::size_t reference_value_size = element.reference_value_size();
-  std::size_t value_size = element.value_size();
-  std::size_t space_dimension = element.space_dimension();
+  // Get elemet
```
Comment (chrisrichardson) [inline cpp/dolfin/function/Function.cpp]: typo

File: cpp/dolfin/function/Function.cpp [hunk 2]
```diff
@@ -266,42 +202,65 @@ void Function::eval(
...
+    const mesh::MeshEntity cell(mesh, tdim, cell_index);
+    restrict(cell, coordinate_dofs, coefficients.data());
+
+    // Compute expansion
+    for (int i = 0; i < space_dimension; ++i)
     {
-      for (std::size_t j = 0; j < value_size; ++j)
+      for (int j = 0; j < value_size; ++j)
       {
         // TODO: Find an Eigen shortcut fot this operation
```
Comment (chrisrichardson) [inline cpp/dolfin/function/Function.cpp]: typo

File: cpp/dolfin/function/Function.h [hunk 3]
```diff
@@ -110,28 +110,18 @@ class Function
...
 
-  /// Evaluate at given point in given cell
+  /// Evaluate the Function at points
   /// @param[in] x The coordinates of the points
-  /// @param[in] cell The cell which contains the given point
-  /// @param[in,out] u The values at the points
+  /// @param[in] cells The index of the cell which contains each point.
+  ///            Negative indices are ignored.
+  /// @param[in,out] u The values at the points. Values are not computed
+  ///                  for points with a negative cel index. This
```
Comment (chrisrichardson) [inline cpp/dolfin/function/Function.h]: typos cel, corrext

File: python/dolfin/function.py [hunk 4]
```diff
@@ -82,11 +81,7 @@ def ufl_evaluate(self, x, component, derivatives):
             # Scalar evaluation
             return self(*x)
 
-    def eval_cell(self, x: np.ndarray, cell, u):
-        u = self._cpp_object.eval_cell(x, cell, u)
-        return u
-
-    def eval(self, x: np.ndarray, bb_tree: cpp.geometry.BoundingBoxTree, u=None) -> np.ndarray:
+    def eval(self, x: np.ndarray, cells: np.ndarray, u=None) -> np.ndarray:
```
Comment (chrisrichardson) [inline python/dolfin/function.py]: Add description of cells to docstring.

**Comment by michalhabera** [comment]
Now you call `element.transform_reference_basis` etc per each point, why is this call not vectorised -- as it way called for all points just once in old code.

**Comment by garth-wells** [comment]
The old code could only process points for one cell per call, which was a performance problem. There are some optimisations that haven't been implemented yet in the case that the points are passed in order by cell.

**Pull request merged**